pkg/util/kernel/find_headers.go

```diff
@@ -0,0 +1,233 @@
+
+		// If no valid directories are found, attempt a fallback to extracting from `/sys/kernel/kheaders.tar.xz`
+		// which is enabled via the `kheaders` kernel module and the `CONFIG_KHEADERS` kernel config option.
+		// The `kheaders` module will be automatically added and removed if present and needed.
+		if dirs, err = getSysfsHeaderDirs(hv); err == nil {
+			return dirs, sysfsHeadersFound, nil
+		}
+		log.Debugf("unable to find system kernel headers: %s", err)
+	}
+
+	dirs = getDownloadedHeaderDirs(headerDownloadDir)
+	if err = validateHeaderDirs(hv, dirs); err == nil {
+		return dirs, downloadedHeadersFound, nil
+	}
+	log.Debugf("unable to find downloaded kernel headers: %s", err)
+
+	if dirs, err = downloadHeaders(headerDownloadDir); err == nil {
+		log.Infof("successfully downloaded kernel headers to %s", dirs)
+		return dirs, downloadSuccess, nil
+	}
+	return nil, downloadFailure, fmt.Errorf("unable to download kernel headers: %w", err)
+}
+
+// validateHeaderDirs verifies that the kernel headers in at least 1 directory matches the kernel version of the running host
+func validateHeaderDirs(hv Version, dirs []string) error {
+	for _, d := range dirs {
+		dirv, err := getHeaderVersion(d)
+		if err != nil {
+			if os.IsNotExist(err) {
+				// if version.h is not found in this directory, keep going
+				continue
+			}
+			return fmt.Errorf("error validating headers version: %w", err)
+		}
+		if dirv != hv {
+			return fmt.Errorf("header version %s does not match host version %s", dirv, hv)
+		}
+		// as long as one directory passes, validate the entire set
+		log.Debugf("found valid kernel headers at %s", d)
+		return nil
+	}
+
+	return fmt.Errorf("no valid kernel headers found")
+}
+
+func getHeaderVersion(path string) (Version, error) {
+	vh := filepath.Join(path, "include/generated/uapi/linux/version.h")
+	f, err := os.Open(vh)
+	if err != nil {
+		return 0, err
+	}
+	defer f.Close()
+	return parseHeaderVersion(f)
+}
+
+func parseHeaderVersion(r io.Reader) (Version, error) {
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		if matches := versionCodeRegexp.FindSubmatch(scanner.Bytes()); matches != nil {
+			code, err := strconv.ParseUint(string(matches[1]), 10, 32)
+			if err != nil {
+				continue
+			}
+			return Version(code), nil
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		return 0, err
+	}
+	return 0, fmt.Errorf("no kernel version found")
+}
+
+func getDefaultHeaderDirs() []string {
+	// KernelVersion == uname -r
+	hi := host.GetStatusInformation()
+	if hi.KernelVersion == "" {
+		return []string{}
+	}
+
+	dirs := []string{
+		fmt.Sprintf(kernelModulesPath, hi.KernelVersion),
+	}
+	switch hi.Platform {
+	case "debian":
+		dirs = append(dirs, fmt.Sprintf(debKernelModulesPath, hi.KernelVersion))
+	case "cos":
+		dirs = append(dirs, fmt.Sprintf(cosKernelModulesPath, hi.KernelVersion))
+	}
+	return dirs
+}
+
+func getDownloadedHeaderDirs(headerDownloadDir string) []string {
```

Comment:
It seems like this function and `getDefaultHeaderDirs` could be combined fairly easily.

Reply:
Good point!